src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/MemoryScope.java

```diff
@@ -254,4 +267,88 @@ void justClose() {
             SCOPED_MEMORY_ACCESS.closeScope(this);
         }
     }
+
+    /**
+     * A functional interface modelling the cleanup action associated with a scope.
+     */
+    interface CleanupAction {
+        void cleanup();
+        CleanupAction dup();
+
+        /** Dummy cleanup action */
+        CleanupAction DUMMY = new CleanupAction() {
+            @Override
+            public void cleanup() {
+                // do nothing
+            }
+
+            @Override
+            public CleanupAction dup() {
+                return this;
+            }
+        };
+    }
+
+    /**
+     * A stateful cleanup action; this action can only be called at most once. The implementation
+     * guarantees this invariant even when multiple threads race to call the {@link #cleanup()} method.
+     */
+    static abstract class BasicCleanupAction implements CleanupAction {
+
+        static final VarHandle CALLED;
+
+        static {
+            try {
+                CALLED = MethodHandles.lookup().findVarHandle(BasicCleanupAction.class, "called", boolean.class);
+            } catch (Throwable ex) {
+                throw new ExceptionInInitializerError(ex);
+            }
+        }
+
+        private boolean called = false;
+
+        abstract void run();
+
+        public final void cleanup() {
+            disable();
+            run();
+        };
+
+        @Override
+        public CleanupAction dup() {
+            disable();
+            return new BasicCleanupAction() {
+                @Override
+                void run() {
+                    BasicCleanupAction.this.run();
```

Comment:
This has the effect of chaining calls, that might accumulate over time. I don't see how to avoid this unless a delegating model is used. But IIUC you want to avoid creating two objects, the "clean up" action as a `Runnable` then the "scope close action" encapsulating the "clean up" action. Perhaps when inline types are available we can do better.

Reply:
I think this is addresses in latest iteration; I used a local class which keeps track of the "root" action, and keeps propagating it. So calls always hit the root action, no chaining.